### ISSUE TITLE
CLI output in CI is less helpful than it used to be

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#1639]
It used to be the case that we printed the diff output during CI, for instance

```
        ~ aws:ecs/service:Service: (update)
            [id=arn:aws:ecs:us-east-1:857628414811:service/cts-auto-web-service-855d766]
            [urn=urn:pulumi:cts-auto::cts::cloud:service:Service$aws:ecs/service:Service::cts-auto-web-service]
          ~ taskDefinition                 : "arn:aws:ecs:us-east-1:857628414811:task-definition/cts-auto-web-service:653" => undefined
        ~ aws:apigateway/restApi:RestApi: (update)
            [id=mqask74dc8]
            [urn=urn:pulumi:cts-auto::cts::cloud:http:API$aws:apigateway/restApi:RestApi::cts-auto-admin-http]
          ~ body                  : "{\"swagger\":\"2.0\",\"info\":{\"title\":\"cts-auto-admin-http\",\"version\":\"1.0\"},\"paths\":{\"/jobs/DefaultData\":{\"post\":{\"x-amazon-apigateway-integration\":{\"passthroughBehavior\":\"when_no_match\",\"httpMethod\":\"POST\",\"type\":\"aws_proxy\",\"uri\":\"arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/arn:aws:lambda:us-east-1:857628414811:function:cts-auto-admin-http4eb2b7c4-5da3716/invocations\"}}},\"/jobs/MasterAccount\":{\"post\":{\"x-amazon-apigateway-integration\":{\"passthroughBehavior\":\"when_no_match\",\"httpMethod\":\"POST\",\"type\":\"aws_proxy\",\"uri\":\"arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/arn:aws:lambda:us-east-1:857628414811:function:cts-auto-admin-http33b66e7b-7ca01ed/invocations\"}}}},\"x-amazon-apigateway-binary-media-types\":[\"*/*\"],\"x-amazon-apigateway-gateway-responses\":{\"MISSING_AUTHENTICATION_TOKEN\":{\"statusCode\":404,\"responseTemplates\":{\"application/json\":\"{\\\"message\\\": \\\"404 Not found\\\" }\"}},\"ACCESS_DENIED\":{\"statusCode\":404,\"responseTemplates\":{\"application/json\":\"{\\\"message\\\": \\\"404 Not found\\\" }\"}}}}" => undefined
        +-aws:apigateway/deployment:Deployment: (replace)
            [id=hwecn0]
            [urn=urn:pulumi:cts-auto::cts::cloud:http:API$aws:apigateway/deployment:Deployment::cts-auto-admin-http]
          ~ description: "Deployment of version 8de6228f" => undefined
          ~ variables  : {
              ~ version: "8de6228f" => undefined
            }
                ~ aws:lambda/function:Function: (update)
                    [id=cts-auto-job-DataSnapshot-cf57930]
                    [urn=urn:pulumi:cts-auto::cts::cloud:timer:Timer$cloud:function:Function$aws:serverless:Function$aws:lambda/function:Function::cts-auto-job-DataSnapshot]
                  ~ code      : archive(assets:0345f95->ada928f) {
                      ~ "__index": asset(text:dafe95b->9241e4a) {
                            ...
                                        });
                                    }
                        }
                    }
```

During the change to the super duper amazing interactive CLI, however, we changed this.  IMHO, the new output is far less useful, particularly because in CI you'll often be trying to debug what happened after the fact.  We can also echo lines for the same resource multiple times, depending on the arrival of certain events, which leads to confusing output.

If we're going to keep the current output, I think we need to make it a little clearer that events are paired (begin/end) somehow.  Alternatively, we could go back to something like the old output which, given that this is our "build log format" in the service already, feels like the right call to me.

**Comment by joeduffy** [comment]
I agree, but having the build log output directly in CI feels right to me.

If not that, maybe something a little more helpful, since the current output is quite opaque.

**Comment by joeduffy** [comment]
Wouldn't this be a better default?

**Comment by joeduffy** [comment]
Even if not, btw, the new output just seems far less useful and hard to follow, especially because of paired events like before/after "same" lines for unchanged resources. I feel like there is definite room for improvement here. I wouldn't drop everything to do something, but I definitely would think (and maybe talk to users) about what would be a useful display here.

**Comment by joeduffy** [comment]
I also wonder if @chrsmith's workflow features influence our thinking here.

**Comment by CyrusNajmabadi** [comment]
> Wouldn't this be a better default?

I personally don't think so.  I don't think "non-interactive" implies "CI".  I would prefer that non-interactive is similar to the normal display, just more line-oriented.  So you hear about things as we hear about them from the engine.

Having the diff display become the default greatly diverges the displays and seems to add a lot of info that i don't think will be generally valuable (as the default).  

>  the new output just seems far less useful and hard to follow, especially because of paired events like before/after "same" lines for unchanged resources.

We could certainly consider trying to do a little tweaking here.  Perhaps buffering events so that if we hear about before/after for something unchanged, (with no intervening messages perhaps) we just display that as a single notification.  

This would at least fit into the mindset of "keep things simple and concise" which also has the benefit of making any real issues actually be more prominent.

**Comment by CyrusNajmabadi** [comment]
> Is our recommendation that anyone using pulumi in CI should pass --diff

I would be fine saying that to someone.  It's unclear to me why that's not a sufficient solution.  THis feels like how things work with many tools.  They have a lot of different choices in terms of their outputs, and people can choose what works best for them for specific environments.

More importantly, it's unclear to me if this is an actual problem people having been encountering and which needs something resolved right now.  On a personal level, i haven't found the CI experience to be problematic, and I end up having to use it continuously across many repos.

> Should we be using our existing CI-sniffing logic in the CLI to default over to --diff by default?

Do any tools really behave that way?  The only reason we sniff CI today is simply because CI's seem like they're interactive, but completely blow up with interactive tools.  

> Do we have any good ideas on how to make the the default enough better that we would not want that in CI?

**Comment by CyrusNajmabadi** [comment]
Closing out.  I think the current state of things is fine.  We have a sensible default that works well in a wide range of scenarios, and we have simple command line switches that customers can use to customize behavior if they would prefer something different.  I'm not seeing a large need to change that.
